src/segment/filter/blocked_bloom/builder.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2024-present, fjall-rs
+// This source code is licensed under both the Apache 2.0 and MIT License
+// (found in the LICENSE-* files in the repository)
+
+use super::{super::bit_array::Builder as BitArrayBuilder, BlockedBloomFilter};
+use crate::segment::filter::{bit_array::BitArrayReader, CACHE_LINE_BYTES};
+
+/// Two hashes that are used for double hashing
+pub type CompositeHash = (u64, u64);
+
+#[derive(Debug, Eq, PartialEq)]
+#[allow(clippy::module_name_repetitions)]
+pub struct Builder {
+    /// Raw bytes exposed as bit array
+    inner: BitArrayBuilder,
+
+    /// Number of hash functions
+    k: usize,
+
+    /// Number of blocks in the blocked bloom filter
+    num_blocks: usize,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl Builder {
+    #[must_use]
+    pub fn build(self) -> BlockedBloomFilter {
+        BlockedBloomFilter {
+            inner: BitArrayReader::new(self.inner.bytes().into()),
+            k: self.k,
+            num_blocks: self.num_blocks,
+        }
+    }
+
+    /// Constructs a bloom filter that can hold `n` items
+    /// while maintaining a certain false positive rate `fpr`.
+    #[must_use]
+    pub fn with_fp_rate(n: usize, fpr: f32) -> Self {
+        use std::f32::consts::LN_2;
+
+        assert!(n > 0);
+
+        // NOTE: Some sensible minimum
+        let fpr = fpr.max(0.000_001);
+
+        // TODO: m and k is still calculated by traditional standard bloom filter formula
+        let m = Self::calculate_m(n, fpr);
+        let bpk = m / n;
+        let k = (((bpk as f32) * LN_2) as usize).max(1);
+
+        let num_blocks = m.div_ceil(CACHE_LINE_BYTES);
+
+        Self {
+            inner: BitArrayBuilder::with_capacity(num_blocks * CACHE_LINE_BYTES),
+            k,
+            num_blocks,
+        }
+    }
+
+    /// Constructs a bloom filter that can hold `n` items
+    /// with `bpk` bits per key.
+    ///
+    /// 10 bits per key is a sensible default.
+    #[must_use]
+    pub fn with_bpk(n: usize, bpk: u8) -> Self {
+        use std::f32::consts::LN_2;
+
+        assert!(bpk > 0);
+        assert!(n > 0);
+
+        let bpk = bpk as usize;
+
+        let m = n * bpk;
+        let k = (((bpk as f32) * LN_2) as usize).max(1);
+
+        let num_blocks = m.div_ceil(CACHE_LINE_BYTES);
```

Comment:
Note that `m` is in *bits*, but `CACHE_LINE_BYTES` is in *bytes*, so currently there are 8x more blocks allocated than necessary.